lib/streamlit/connections/connection_factory.py

```diff
@@ -0,0 +1,131 @@
+# BaseConnection are responsible for binding the type parameter of BaseConnection to a
+# concrete type, but the type it was bound to isn't important to us here.
+ConnectionClass = TypeVar("ConnectionClass", bound=BaseConnection[Any])
+
+
+# NOTE: Adding support for a new first party connection requires:
+#   1. Adding the new connection name and class to this function.
+#   2. Writing a new @overload signature mapping the connection's name to its class.
+# TODO(vdonato): Some way to test that optional dependencies required by a connection
+# don't cause `ModuleNotFoundError`s until the connection is actually instantiated.
+def _get_first_party_connection(connection_name: str):
+    FIRST_PARTY_CONNECTIONS = {"snowpark", "sql", "files", "s3", "gcs"}
+
+    if connection_name == "sql":
+        return SQL
+    elif connection_name == "snowpark":
+        return Snowpark
+    elif connection_name == "files":
+        return FileSystem
+    elif connection_name == "s3":
+        return FileSystem
+    elif connection_name == "gcs":
+        return FileSystem
+
+    raise StreamlitAPIException(
+        f"Invalid connection {connection_name}. Supported connection classes: {FIRST_PARTY_CONNECTIONS}"
+    )
+
+
+def _validate(conn: ConnectionClass) -> bool:
+    return conn.is_connected()
+
+
+@overload
+def connection(
+    connection_class: Literal["sql"], name: str = "default", **kwargs
+) -> "SQL":
+    ...
+
+
+@overload
+def connection(
+    connection_class: Literal["snowpark"], name: str = "default", **kwargs
+) -> "Snowpark":
+    ...
+
+
+@overload
+def connection(
+    connection_class: Literal["files"], name: str = "default", **kwargs
+) -> "FileSystem":
+    ...
+
+
+@overload
+def connection(
+    connection_class: Literal["s3"], name: str = "default", **kwargs
+) -> "FileSystem":
+    ...
+
+
+@overload
+def connection(
+    connection_class: Literal["gcs"], name: str = "default", **kwargs
+) -> "FileSystem":
+    ...
+
+
+@overload
+def connection(
+    connection_class: Type[ConnectionClass], name: str = "default", **kwargs
+) -> ConnectionClass:
+    ...
+
+
+# TODO(vdonato): Write a docstring for this function.
+# TODO(vdonato): Maybe support st.connection.sql() syntax as an alias for
+#                st.connection("sql") if we decide we want to do that.
+@gather_metrics("connection")
+@cache_resource(validate=_validate)
+def connection(connection_class, name="default", **kwargs):
+    if type(connection_class) == str:
+        if connection_class == "s3":
+            kwargs["protocol"] = kwargs.get("protocol", "s3")
+            if name == "default":
+                name = "s3"
+        if connection_class == "gcs":
+            kwargs["protocol"] = kwargs.get("protocol", "gcs")
+            if name == "default":
+                name = "gcs"
+        connection_class = _get_first_party_connection(connection_class)
+
+    try:
+        return connection_class(
```

Comment:
@vdonato For the sake of metrics, could we separate out just this part of the function, and wrap that in @gather_metrics? That way we would always have an actual Connection Class, so we can tell what kind of connection it is, rather than sometimes having a class, and sometimes having a string.

cc @sfc-gh-jcarroll

Reply:
Ack -- I'm just starting the productionization of this code into what we'll be merging into `develop`, so will do this in the new feature branch being created